### PR TITLE
[4.x] Support custom blueprint namespaces

### PR DESCRIPTION
**Pull request by ryanmitchell**
Taken from https://github.com/statamic/ideas/issues/644

I've added support for adding custom blueprint namespaces:

```php
\Statamic\Facades\Blueprint::addNamespace('custom', 'path/to/blueprints');
```

I've also made it possible to edit these through the CP. 
(Duncan wants this ability for Runway).

### CODE REVIEW

**Comment by edalzell** [comment]
oh very nice!

**Comment by bretterer** [comment]
This is a much needed addition for addons.  I can not wait to see this in core for the plugin I am developing

**Comment by bretterer** [comment]
Any thoughts on getting this into a release soon?

**Comment by edalzell** [comment]
> Any thoughts on getting this into a release soon?

You can always composer patch it in.

**Comment by ryanmitchell** [comment]
@bretterer it will be reviewed once the core team have time to look at it. It may not be a priority for them, even if it is for you. You can use composer patches to pull in any PR thats not yet merged.

**Review by duncanmcclean — CHANGES_REQUESTED** [review]
This PR works great, thanks! Just a few things I spotted while reviewing the code...

File: src/Fields/BlueprintRepository.php
Comment (jasonvarga) [reply src/Fields/BlueprintRepository.php:208]: If you save blueprints that come from a custom namespace in the CP, they get saved into the vendor directory. The original ones intentionally don't get modified.

Comment (ryanmitchell) [reply src/Fields/BlueprintRepository.php:208]: Yeah this was intentional, it's the same as how additionally namespaced fieldsets work. I looked at it like you are modifying the supplied blueprint, so it should be vendor folder-ed, similar to modifying a view from a package.

File: resources/views/blueprints/index.blade.php
Comment (ryanmitchell) [reply resources/views/blueprints/index.blade.php:190]: A generic one might be simpler as otherwise you'd either need to add an icon() method to the blueprint class or add an icon when adding a namespace so all blueprints in the namespace have the same icon. 

Do we just run with https://github.com/statamic/cms/blob/4.x/resources/svg/icons/light/blueprint.svg for now?

**Comment by ryanmitchell** [comment]
I didnt add the ability to create blueprints (I thought that would be unnecessary but happy to be overruled). I guess for the Runway use case you could provide a command that creates a new blueprint in the namespace?

File: src/Fields/BlueprintRepository.php
Comment (duncanmcclean) [reply src/Fields/BlueprintRepository.php:208]: Okay! That makes sense. 👍

File: resources/views/blueprints/index.blade.php
Comment (duncanmcclean) [reply resources/views/blueprints/index.blade.php:190]: Yeah, I think that might be the best one to use.

**Comment by duncanmcclean** [comment]
> I didnt add the ability to create blueprints (I thought that would be unnecessary but happy to be overruled). I guess for the Runway use case you could provide a command that creates a new blueprint in the namespace? 

Yeah, Runway doesn't need it. It'll just create blueprints for each resource if one doesn't already exist.

Not sure if it's something that might be useful for non-Runway use cases though 🤔

File: resources/views/blueprints/index.blade.php
Comment (ryanmitchell) [reply resources/views/blueprints/index.blade.php]: Done

**Comment by ryanmitchell** [comment]
> Not sure if it's something that might be useful for non-Runway use cases though 🤔

If someone needs it we can always come back to it later

**Comment by ryanmitchell** [comment]
When registering an additional namespace would it make sense to also register a permission 'configure {namespace} fields` which we then check to ensure the user can modify those fields.

It would mean reconfiguring the blueprint/index a little bit but might be worth it to give people flexibility?

**Comment by duncanmcclean** [comment]
> When registering an additional namespace would it make sense to also register a permission 'configure {namespace} fields` which we then check to ensure the user can modify those fields.
> 
> It would mean reconfiguring the blueprint/index a little bit but might be worth it to give people flexibility?

I don't *think* we need the permissions to be separate. We don't have separate permissions for collections/globals/etc, apart from forms, which are a bit of a special case.

**Review by duncanmcclean — APPROVED** [review]
This PR is working great! I've got a draft PR on the Runway side to implement this: duncanmcclean/runway#373

I'll leave this open for @jasonvarga to review before merging.

**Comment by martinolson** [comment]
Awesome work, looking forward to this!

Thank you!

File: src/Fields/Blueprint.php
Comment (jasonvarga) [inline src/Fields/Blueprint.php:113]: Do you remember why this was needed? 

The initial path is supposed to just be where the file actually was.
e.g. If it was in the namespaced directory, it'll be that. If it was customized by a user, it'll be the vendor path.

Everything seems to work fine if you remove this condition.

**Comment by edalzell** [comment]
WOO HOO

**Comment by ryanmitchell** [comment]
Thanks!

**Pull request merged**
